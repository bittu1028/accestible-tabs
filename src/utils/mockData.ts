
import { ITabsData } from '../components/Tabs.type';

export const tabsConfig:ITabsData[] = [
    {
      label: "Tab1",
      content: "Content Panel 1",
      isDisabled: false,
    },
    {
      label: "Tab2",
      content: "Content Panel 2",
      isDisabled: true,
    },
    {
      label: "Tab3",
      content: "Content Panel 3",
      isDisabled: false,
    }, {
      label: "Tab4",
      content: "Content Panel 4",
      isDisabled: false,
    }, {
      label: "Tab5",
      content: "Content Panel 5",
      isDisabled: false,
    },
  ];