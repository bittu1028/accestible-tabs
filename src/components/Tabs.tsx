import { useEffect, useRef, useState } from "react";
import { Tab, TabItem, TabList, TabPanel, Wrap } from "./StyledTab";
import { getNextTab, getPrevTab } from "../utils/tabUtils";
import { ITabs, ITabsData } from "./Tabs.type";

const Tabs = ({ tabs, id }: ITabs) => {
  const [activeTab, setActiveTab] = useState<string>(tabs[0].label);

  // Store the tab element references here
  const tabRefs: any = useRef({});

  useEffect(() => {
    focusTab(tabs[0]);
  }, [tabs])

  const handleClick = (tab: ITabsData) => (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    setActiveTab(tab.label);
  };

  const getId = (...rest: any) => [...rest, id].join("-");

  const focusTab = (tab: ITabsData) => {
    setActiveTab(tab.label);
    // Get the reference from the `tabRefs` and actual move focus to the tab
    tabRefs.current[tab.label].focus();
  };

  //key board navigation
  const handleKeyboard = (currentTabIndex: number) => (event: any) => {
    const tabCount = tabs.length;
    const firstTab = tabs[0];
    const lastTab = tabs[tabCount - 1];
    const nextTab = tabs[getNextTab(currentTabIndex + 1, tabs)];
    const prevTab = tabs[getPrevTab(currentTabIndex - 1, tabs)];

    if (event.key === "ArrowRight") {
      if (tabCount > currentTabIndex + 1) return focusTab(nextTab);
      return focusTab(firstTab);
    }

    if (event.key === "ArrowLeft") {
      if (currentTabIndex > 0) return focusTab(prevTab);
      return focusTab(lastTab);
    }

    if (event.key === "Home") {
      event.preventDefault();
      return focusTab(firstTab);
    }

    if (event.key === "End") {
      event.preventDefault();
      return focusTab(lastTab);
    }
  };

  return (
    <Wrap>
      <TabList aria-orientation="horizontal" role="tablist">
        {tabs.map((tab, index) => {
          const isActive = activeTab === tab.label;

          return (
            <TabItem isDisabled={tab.isDisabled} key={tab.label}>
              <Tab
                aria-selected={isActive}
                aria-disabled={tab.isDisabled}
                id={getId(tab.label)}
                aria-controls={getId(tab.label, "tabpanel")}
                isActive={isActive}
                isDisabled={tab.isDisabled}
                disabled={tab.isDisabled}
                onClick={handleClick(tab)}
                onKeyDown={handleKeyboard(index)}
                tabIndex={isActive ? 0 : -1}
                ref={(ref) => (tabRefs.current[tab.label] = ref)}
                role="tab"
              >
                {tab.label}
              </Tab>
            </TabItem>
          );
        })}
      </TabList>
      {tabs.map((tab) => (
        <TabPanel
          aria-labelledby={getId(tab.label)}
          id={getId(tab.label, "tabpanel")}
          hidden={activeTab !== tab.label}
          key={tab.label}
          tabIndex={0}
          role="tabpanel"
        >
          {tab.content}
        </TabPanel>
      ))}
    </Wrap>
  );
};

export default Tabs;
