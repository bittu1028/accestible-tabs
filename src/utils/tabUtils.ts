import { ITabsData } from "../components/Tabs.type";


export const isTabDisabled = (tab:ITabsData) => {
    return tab && tab.isDisabled;
}

export const getTab = (index: number, tabs: ITabsData[]) => {
    return tabs[index];
}

export const getNextTab =(index:number, tabs: ITabsData[]) => {
  // Look for non-disabled tab from index to the last tab on the right;
  const count  = tabs.length;
  for (let i = index; i < count; i++) {
    const nextTab = getTab(i, tabs);
    if (!isTabDisabled(nextTab)) {
      return i;
    }
  }
  return index;
}

export const getPrevTab =(index:number, tabs: ITabsData[]) => {
    // Look for non-disabled tab from index to the first tab on the left
    for (let i = index; i >= 0; i--) {
      const prevTab = getTab(i, tabs);
      if (!isTabDisabled(prevTab)) {
        return i;
      }
    }
    return index;
}