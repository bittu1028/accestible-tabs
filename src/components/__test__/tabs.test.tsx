/* eslint-disable testing-library/no-node-access */
import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import Tabs from "../Tabs";

import { tabsConfig } from "../../utils/mockData";

describe("Tabs Component", () => {
  test("renders learn react link", () => {
    const { asFragment } = render(<Tabs tabs={tabsConfig} />);
    expect(asFragment).toMatchSnapshot();
  });

  test("expect tab1 to be active and selected", () => {
    render(<Tabs tabs={tabsConfig} />);
    const node = screen.getByRole("tab", { selected: true });
    expect(node).toHaveTextContent("Tab1");
  });

  test("expect tab2 to be disabled", () => {
    render(<Tabs tabs={tabsConfig} />);
    expect(screen.getByText(/Tab2/i).closest("button")).toBeDisabled();
  });

  test("expect arrowRight key to trigger next tab in the list and not select disabled tab Tab2", () => {
    render(<Tabs tabs={tabsConfig} />);
    const button = screen.getByRole("tab", { selected: true });
    expect(button).toHaveTextContent("Tab1");
    fireEvent.keyDown(button, { key: "ArrowRight", code: "KeyArrowRight" });
    const nextButton = screen.getByRole("tab", { selected: true });
    expect(nextButton).toHaveTextContent("Tab3");
  });

  test("expect arrowLeft key to trigger prev tab in the list", () => {
    render(<Tabs tabs={tabsConfig} />);
    const button = screen.getByRole("tab", { selected: true });
    expect(button).toHaveTextContent("Tab1");
    fireEvent.keyDown(button, { key: "ArrowLeft", code: "KeyArrowLeft" });
    const nextButton = screen.getByRole("tab", { selected: true });
    expect(nextButton).toHaveTextContent("Tab5");
  });

  test("expect Home key to trigger first tab in the list", () => {
    render(<Tabs tabs={tabsConfig} />);
    const button = screen.getByRole("tab", { selected: true });
    expect(button).toHaveTextContent("Tab1");
    fireEvent.keyDown(button, { key: "ArrowRight", code: "KeyArrowRight" });
    fireEvent.keyDown(button, { key: "Home", code: "KeyHome" });
    const nextButton = screen.getByRole("tab", { selected: true });
    expect(nextButton).toHaveTextContent("Tab1");
  });

  test("expect End key to trigger last tab in the list", () => {
    render(<Tabs tabs={tabsConfig} />);
    const button = screen.getByRole("tab", { selected: true });
    expect(button).toHaveTextContent("Tab1");
    fireEvent.keyDown(button, { key: "End", code: "KeyArrowEnd" });
    const nextButton = screen.getByRole("tab", { selected: true });
    expect(nextButton).toHaveTextContent("Tab5");
  });

});
