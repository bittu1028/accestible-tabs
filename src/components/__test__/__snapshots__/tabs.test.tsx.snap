// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component renders learn react link 1`] = `[Function]`;
